src/kern/arm/bsp/drivers/mru.cpp

```diff
@@ -0,0 +1,246 @@
+    Ch_cfg0   = 0x00,
+    Ch_cfg1   = 0x04,
+    Ch_mbstat = 0x08,
+  };
+
+  struct Cfg0
+  {
+    Unsigned32 raw;
+    explicit Cfg0(Unsigned32 v) : raw(v) {}
+    CXX_BITFIELD_MEMBER( 0,  0, che, raw);
+    CXX_BITFIELD_MEMBER( 1,  1, chr, raw);
+    CXX_BITFIELD_MEMBER( 2,  2, ie, raw);
+    CXX_BITFIELD_MEMBER(16, 23, mbe, raw);
+  };
+
+  // Extend Irq_chip::Mode. Use additional mode bits for MRU channel set_mode()
+  // calls.
+  struct Mru_ch_mode : Mode
+  {
+    using Mode::Mode;
+    Mru_ch_mode(Mode const &m) : Mode(m.raw) {}
+
+    CXX_BITFIELD_MEMBER(15, 15, set_mb_enable, raw);
+    CXX_BITFIELD_MEMBER(16, 23, mbe, raw);
+  };
+
+  class Mru_ppi : public Irq_base
+  {
+  public:
+    Mru_ppi(Mru *mru)
+    : _mru(mru)
+    {
+      set_hit(handler_wrapper<Mru_ppi>);
+    }
+
+    void init(Gic *parent, unsigned id)
+    {
+      check(parent->attach(this, id, false));
+      chip()->set_mode_percpu(current_cpu(), pin(),
+                              Irq_chip::Mode::F_level_high);
+      chip()->unmask_percpu(current_cpu(), pin());
+    }
+
+    void handle(Upstream_irq const *ui)
+    {
+      // Acknowledge before dispatching. The handle_multi_pending() method will
+      // take the shortcut to user space.
+      chip()->ack(pin());
+      Upstream_irq::ack(ui);
+      _mru->handle_multi_pending<Mru>(ui);
+    }
+
+  private:
+    void switch_mode(bool) override {}
+
+    Mru *_mru;
+  };
+
+  inline void cfg0(unsigned ch, Cfg0 v)
+  { write<Unsigned32>(v.raw, ch * Ch_offset); }
+
+  inline Cfg0 cfg0(unsigned ch) const
+  { return Cfg0(read<Unsigned32>(ch * Ch_offset)); }
+};
+
+// ------------------------------------------------------------------------
+INTERFACE [arm && pic_gic && cpu_virt]:
+
+EXTENSION class Mru
+{
+  // Helper to temporarily set VSCTLR.VMID to 0. Needed to access peripherals
+  // from the kernel in case a VM is interrupted.
+  class Vmid_guard
+  {
+    unsigned _vsctlr;
+
+  public:
+    Vmid_guard()
+    {
+      asm volatile ("mrc p15, 4, %0, c2, c0, 0" : "=r"(_vsctlr));
+      asm volatile (
+          "mcr p15, 4, %0, c2, c0, 0   \n" // VSCTLR
+          "isb                         \n"
+          : : "r"(0));
+    }
+
+    ~Vmid_guard()
+    {
+      asm volatile (
+          "mcr p15, 4, %0, c2, c0, 0   \n" // VSCTLR
+          : : "r"(_vsctlr));
+    }
+  };
+};
+
+// ------------------------------------------------------------------------
+INTERFACE [arm && pic_gic && !cpu_virt]:
+
+EXTENSION class Mru
+{
+  // When running on EL1, the MRU is supposed to be accessible...
+  class Vmid_guard {
+  public:
+    Vmid_guard() { /* dummy to prevent warning */ }
+    ~Vmid_guard() {}
+  };
+};
+
+// ------------------------------------------------------------------------
+IMPLEMENTATION [arm && pic_gic]:
+
+#include "irq_mgr_multi_chip.h"
+#include "platform_control.h"
+#include "kip.h"
+#include "kmem_mmio.h"
+#include "cpu.h"
+
+PRIVATE
+void
+Mru::init()
+{
+  for (unsigned i = 0; i < Nr_irqs; i++)
```

Comment:
I would pass `Nr_irqs` as parameter to `init()` because the constant is not defined in this file.